### ISSUE TITLE
Start Application

### ISSUE THREAD

**Issue by sunami09** [esnet/sense-rtmon#72]
# Troubleshooting `start.sh` Script Execution

## Issue Description

During the execution of `start.sh`, a script designed to initialize various services including Grafana, Prometheus, and Pushgateway, an error was encountered related to the absence of the `lsof` command on the system.

### Error Output

The script checks if ports 3000, 9090, 9091, and 9469 are in use by attempting to use `lsof` to list processes listening on these ports. However, the execution of these commands resulted in errors indicating that `lsof` was not found:

```plaintext
!!    Make sure Port 3000, 9090, 9091, 9469 are not in use
!!    sudo lsof -i -P -n | grep LISTEN
!!    Check Port 3000
sudo: lsof: command not found
!!    Check Port 9090
sudo: lsof: command not found
!!    Check Port 9091
sudo: lsof: command not found
!!    Check Port 9469
sudo: lsof: command not found
```

Despite these errors, the script proceeded with the configuration parsing and successfully deployed the Docker stack for the required services.

## Resolution

The `lsof` (LiSt Open Files) command is a utility to list information about files opened by processes. It's not installed by default on all Linux distributions. To resolve the issue, you need to install `lsof` on your system.

### Installing `lsof`

On RHEL (Red Hat Enterprise Linux) and similar systems, you can install `lsof` using `yum`:

```bash
sudo yum install lsof
```

### Verification

After installing `lsof`, rerun the relevant part of the script or the entire script to ensure it can now successfully check if the specified ports are in use without encountering the previous errors.

## Script Execution

Post-resolution, the `start.sh` script was executed again. This time, the script ran without encountering the errors related to `lsof`. The script output confirmed the successful deployment of the services and provided instructions for next steps, including Grafana dashboard setup and data source configuration.

### Successful Script Execution Highlights

- Ports were successfully checked for availability.
- Configuration file was parsed without issues.
- Docker stack for Grafana, Prometheus, and Pushgateway was deployed.
- Users were instructed on how to manually set up the Grafana data source and authorization API key.

## Conclusion

The initial execution of `start.sh` revealed a dependency on the `lsof` utility, which was not met. Installing `lsof` resolved the issue, allowing for the successful execution of the script. This emphasizes the importance of ensuring all script dependencies are met to facilitate smooth operations and deployments.

**Comment by sunami09** [comment]
We can add this in the `install.sh` script 
```bash
sudo yum install lsof
```

**Comment by sunami09** [comment]
# Docker Services Verification and Web Interface Accessibility

After initiating the Docker stack for our monitoring and visualization tools, we verified that all the necessary Docker containers are up and running. Additionally, we confirmed the accessibility of their respective web interfaces.

## Docker Containers Status

The following command was executed to list all running Docker containers:

```bash
docker ps
```

**Output:**

The output confirmed the successful deployment and running status of Grafana, Prometheus, and PushGateway services:

```plaintext
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS      NAMES
a6594b128369   grafana/grafana-enterprise:latest     "/run.sh"                2 seconds ago    Up 1 second     3000/tcp   cloud_grafana.1.15xs99qouuefv2jocg46oc5zr
aec0420405e6   prom/prometheus:v2.2.1                "/bin/prometheus --c…"   6 seconds ago    Up 5 seconds    9090/tcp   cloud_prometheus.1.jc7vnnxy8qnu3zi27rwqj8uj7
fb85d999005c   prom/pushgateway:latest               "/bin/pushgateway"       10 seconds ago   Up 6 seconds    9091/tcp   cloud_pushgateway.1.pwp0bnswm8juynju7ojfk9ofn
e9a1c649fa5d   ricoberger/script_exporter:v2.16.0    "/sbin/tini -- /bin/…"   14 seconds ago   Up 12 seconds   9469/tcp   cloud_script_exporter.1.xrhm1k1n4uflxix6xcta6al9r
```

## Web Interface Accessibility

Following the container deployment, we accessed the web interfaces for Grafana, Prometheus, and PushGateway to ensure they were correctly set up and accessible.

### Grafana Dashboard

Accessible at `http://100.26.56.42:3000/`, the Grafana dashboard is the central visualization tool for our metrics:

![Grafana Dashboard](https://github.com/esnet/sense-rtmon/assets/66564001/b0dcfddc-49e1-43a1-b491-a126bfc1dd5f)

### Prometheus Targets

The Prometheus targets page, accessible at `http://100.26.56.42:9090/targets`, shows the metrics sources being monitored:

![Prometheus Targets](https://github.com/esnet/sense-rtmon/assets/66564001/bd6b8340-2c05-41e4-a7ad-6f8c6b0465aa)

### PushGateway Interface

PushGateway's web interface, available at `http://100.26.56.42:9091/`, allows for pushing metrics from jobs that cannot be scraped:

![PushGateway WebPage](https://github.com/esnet/sense-rtmon/assets/66564001/c168296b-b2c2-42b8-a95e-117f529bda30)

## Conclusion

The successful verification of the Docker containers and the accessibility of the Grafana, Prometheus, and PushGateway web interfaces confirm the correct deployment and configuration of our monitoring stack. This setup enables comprehensive monitoring and visualization capabilities for our system metrics.
